mesonbuild/backend/ninjabackend.py

```diff
@@ -1911,6 +1911,33 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
                 # Rust uses -l for non rust dependencies, but we still need to
                 # add dylib=foo
                 args += ['-l', f'dylib={d.name}']
+
+        # Since 1.61.0 Rust has a special modifier for whole-archive linking,
+        # before that it would treat linking two static libraries as
+        # whole-arachive linking.
+        if mesonlib.version_compare(rustc.version, '>= 1.61.0'):
+            whole_archive = ':+whole-archive'
+        else:
+            whole_archive = ''
+
+        for d in target.link_whole_targets:
+            linkdirs.add(d.subdir)
+            if d.uses_rust():
+                # specify `extern CRATE_NAME=OUTPUT_FILE` for each Rust
+                # dependency, so that collisions with libraries in rustc's
+                # sysroot don't cause ambiguity
+                args += ['--extern', '{}={}'.format(d.name, os.path.join(d.subdir, d.filename))]
+                project_deps.append(RustDep(d.name, self.rust_crates[d.name].order))
+            else:
+                # Rustc doesn't follow Meson's convention that static libraries
+                # are called .a, and import libraries are .lib, so we have to
+                # manually handle that.
+                if rustc.linker.id in {'link', 'lld-link'}:
+                    link_whole_arg = rustc.linker.get_link_whole_for([self.get_target_filename_for_linking(d)])
+                    args += ['-C', f'link-arg={link_whole_arg}']
```

Comment:
Shouldn't this be something like:

```python
for arg in rustc.linker.get_link_whole_for(...):
    args += ['-C', f'link-arg={arg}']
```

Otherwise you will get: `['-C', "link-arg=['-force_load', 'libfoo.a']"]`? Codecov says this whole block is not covered at all, that seems wrong. But maybe this codepath is indeed not being hit. Can the existing test be modified to fix that?

Reply:
Actually, now that I think about this... The current behavior is that *all* targets are whole-archive linked (before Rustc 1.61, I mean). So... Is this going to be a backwards incompatible change, or a change in the implementation deals of rustc?  @eli-schwartz, @nirbheek, Opinions?

Reply:
I don't think it's our job to maintain backwards-compatibility for rustc. They chose to change it, and we should just follow it. Especially when we have probably single-digit users of our rust backend right now.